# Capstone - Final Project
## Author - Salinee Kingbaisomboon
### UW NetID: 1950831

Milestone 3 continues your work with the diaper manufacturing problem, using the same datasets:
1. A dataset file SECOM containing 1567 examples, each with 591 features, presented in a 1567 x 591 matrix
2. A labels file listing the classifications and date time stamp for each example 

## Instructions
1. Build a simple neural networks model
2. Build a DNN model
3. Build a RNN model
4. Summarize your findings with examples.  Explain what the manufacturer should focus on to optimize the diaper manufacturing process
5. Solicit specific feedback on your code

In [3]:
# Load necessary libraries
import pandas as pd
import numpy as np

from collections import Counter

from imblearn.over_sampling import SMOTE

from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import RFE # Recursive Feature Elimination (for backward model selection)
from sklearn.neural_network import MLPClassifier

import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, SimpleRNN, Embedding, GRU

import warnings

warnings.filterwarnings("ignore") # To suppress warning

pd.options.display.max_rows = None
pd.options.display.max_columns = None

# Declare Functions used in this milestone

In [4]:
# Function to replace missing numeric values
def replace_missing_value(x, col):
    # We determine the locations of NaN
    HasNan = np.isnan(x.loc[:, col]) 
    # If there are NaN
    if sum(HasNan) > 0:
        # Calculate the Median for current column without Nan
        Median = np.nanmedian(x.loc[:, col])
        # Replace the missing value with Median
        x.loc[HasNan, col] = Median
    return x.loc[:, col]

In [5]:
# function that returns a tensorflow.data.Dataset object
def input_fn(features, labels, training=True, batch_size=256):
    """An input function for training or evaluating"""
    # Convert the inputs to a Dataset.
    dataset = tf.data.Dataset.from_tensor_slices((dict(features), labels))

    # Shuffle and repeat if you are in training mode.
    if training:
        dataset = dataset.shuffle(1000).repeat()
    
    return dataset.batch(batch_size)

# Read, merge and perform data cleaning

In [6]:
# Load data and label
dataFilename = 'https://archive.ics.uci.edu/ml/machine-learning-databases/secom/secom.data'
data_df = pd.read_csv(dataFilename, header=None, delim_whitespace=True)

labelFilename = 'https://archive.ics.uci.edu/ml/machine-learning-databases/secom/secom_labels.data'
label_df = pd.read_csv(labelFilename, header=None, delim_whitespace=True)

In [7]:
# Set pseudo names for input data
for col in data_df.columns:
    data_df.rename(columns = {col:'feature_'+ str(col)}, inplace = True)
    
# Set pseudo name for label data
label_df.rename(columns = {0:'HouseLineTestingResult'}, inplace = True)
label_df.rename(columns = {1:'TimeStamp'}, inplace = True)

# Since we didn't do the time series analysis, we don't need the timestamp column
# Drop the unused columns
label_df.drop(['TimeStamp'],axis=1, inplace=True)

In [8]:
# Clean input data
for col in data_df.columns:
    # Replace missing values with Median for current column without Nan
    data_df.loc[:, col] = replace_missing_value(data_df, col)

In [9]:
# Merge input and label data into one
df = pd.concat([data_df,label_df],axis=1)

In [10]:
# Normalize data frame
cols = df.columns
min_max_scaler = preprocessing.MinMaxScaler()
np_scaled = min_max_scaler.fit_transform(df)
df_normalized = pd.DataFrame(np_scaled, columns = cols)

## Explore Data

In [11]:
# View first 5 rows
df_normalized.head()

feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
0   0.469231   0.589292   0.499096   0.379841   0.000609        0.0   
1   0.575003   0.445535   0.666763   0.393982   0.000133        0.0   
2   0.308868   0.583388   0.493903   0.457065   0.000744        0.0   
3   0.400385   0.466998   0.543479   0.244894   0.000574        0.0   
4   0.471367   0.500400   0.678327   0.357067   0.000765        0.0   

   feature_6  feature_7  feature_8  feature_9  feature_10  feature_11  \
0   0.328562   0.965785   0.665019   0.542479    0.358362    0.880692   
1   0.428942   0.969673   0.656639   0.412315    0.228669    0.932908   
2   0.283455   0.965008   0.542759   0.448168    0.411832    0.929265   
3   0.469123   0.946345   0.638590   0.319564    0.359499    0.933515   
4   0.387631   0.960342   0.670606   0.392050    0.315131    0.915301   

   feature_12  feature_13  feature_14  feature_15  feature_16  feature_17  \
0    0.226185         0.0    0.329909    0.165668    0.056644    0.958559   
1    0.205145         0.0    0.457039    0.165391    0.048683    0.963740   
2    0.221497         0.0    0.420091    0.169405    0.049237    0.957079   
3    0.219610         0.0    0.425265    0.180762    0.053078    0.960286   
4    0.220658         0.0    0.480818    0.177310    0.059646    0.972126   

   feature_18  feature_19  feature_20  feature_21  feature_22  feature_23  \
0    0.500188    0.848739    0.814395    0.242124    0.797675    0.481224   
1    0.476296    0.830040    0.740957    0.238978    0.712274    0.525355   
2    0.506806    0.855614    0.849836    0.238104    0.738940    0.480961   
3    0.495005    0.835631    0.808915    0.235237    0.724308    0.443066   
4    0.483114    0.834120    0.763975    0.234118    0.720752    0.485779   

   feature_24  feature_25  feature_26  feature_27  feature_28  feature_29  \
0    0.538057    0.647599    0.863698    0.398104    0.261259    0.476550   
1    0.455345    0.938169    0.981245    0.964903    0.487686    0.562509   
2    0.445790    0.948944    0.988650    0.989554    0.418016    0.585923   
3    0.454757    0.950029    0.976130    0.955045    0.190989    0.695296   
4    0.516127    0.931950    0.969992    0.949862    0.185584    0.875018   

   feature_30  feature_31  feature_32  feature_33  feature_34  feature_35  \
0    0.514343    0.529986    0.009554    0.121293    0.078721    0.019008   
1    0.701594    0.492686    0.076817    0.139530    0.083009    0.019793   
2    0.554980    0.559058    0.070202    0.067069    0.032024    0.014189   
3    0.689243    0.479595    0.077053    0.068333    0.067953    0.014225   
4    0.938247    0.439918    0.140226    0.073974    0.041585    0.015487   

   feature_36  feature_37  feature_38  feature_39  feature_40  feature_41  \
0    0.921279    0.458921    0.603022    0.267583    0.706799    0.120954   
1    0.916991    0.000000    0.757277    0.295521    0.907068    0.075058   
2    0.967976    0.302570    0.000000    0.318310    0.152753    0.145166   
3    0.932047    0.090745    0.530100    0.245947    0.891127    0.035697   
4    0.958415    0.406311    0.383630    0.448568    0.885105    0.060199   

   feature_42  feature_43  feature_44  feature_45  feature_46  feature_47  \
0         0.0    0.288419    0.453178    0.319010    0.261238    0.798994   
1         0.0    0.274732    0.360793    0.360565    0.278107    0.780757   
2         0.0    0.625997    0.260730    0.340259    0.388916    0.791108   
3         0.0    0.586889    0.293581    0.340259    0.378747    0.794558   
4         0.0    0.306446    0.594777    1.000000    1.000000    0.529771   

   feature_48  feature_49  feature_50  feature_51  feature_52  feature_53  \
0    0.411986         0.0    0.280399    0.815716         0.0    0.837429   
1    0.535243         0.0    0.395517    0.755483         0.0    0.835539   
2    0.408149         0.0    0.495000    0.666551         0.0    0.737240   
3    0.447084         0.0    0.496776    0.685566  

In [12]:
print(df_normalized.shape) # Print DataFrame's size
print(df_normalized.dtypes) # Print DataFrame's data types

(1567, 591)
feature_0                 float64
feature_1                 float64
feature_2                 float64
feature_3                 float64
feature_4                 float64
feature_5                 float64
feature_6                 float64
feature_7                 float64
feature_8                 float64
feature_9                 float64
feature_10                float64
feature_11                float64
feature_12                float64
feature_13                float64
feature_14                float64
feature_15                float64
feature_16                float64
feature_17                float64
feature_18                float64
feature_19                float64
feature_20                float64
feature_21                float64
feature_22                float64
feature_23                float64
feature_24                float64
feature_25                float64
feature_26                float64
feature_27                float64
feature_28                float64
fe

In [13]:
# View Statistic Summary
df_normalized.describe()

feature_0    feature_1    feature_2    feature_3    feature_4  \
count  1567.000000  1567.000000  1567.000000  1567.000000  1567.000000   
mean      0.442338     0.490215     0.549443     0.375604     0.003133   
std       0.119849     0.116663     0.115397     0.118394     0.050369   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.364413     0.427715     0.473043     0.291756     0.000302   
50%       0.437523     0.495361     0.551465     0.345949     0.000570   
75%       0.511001     0.552567     0.618191     0.428036     0.000752   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

       feature_5    feature_6    feature_7    feature_8    feature_9  \
count     1567.0  1567.000000  1567.000000  1567.000000  1567.000000   
mean         0.0     0.402906     0.947316     0.584143     0.409652   
std          0.0     0.131775     0.069484     0.158679     0.117744   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.335448     0.941680     0.473249     0.332034   
50%          0.0     0.411304     0.951788     0.581435     0.406080   
75%          0.0     0.475348     0.962675     0.700150     0.481684   
max          0.0     1.000000     1.000000     1.000000     1.000000   

        feature_10   feature_11   feature_12  feature_13   feature_14  \
count  1567.000000  1567.000000  1567.000000      1567.0  1567.000000   
mean      0.398704     0.937931     0.198578         0.0     0.390583   
std       0.105758     0.037778     0.036189         0.0     0.161524   
min       0.000000     0.000000     0.000000         0.0     0.000000   
25%       0.333333     0.918944     0.178285         0.0     0.280245   
50%       0.401593     0.942319     0.193901         0.0     0.388369   
75%       0.464164     0.959016     0.221373         0.0     0.497734   
max       1.000000     1.000000     1.000000         0.0     1.000000   

        feature_15   feature_16   feature_17   feature_18   feature_19  \
count  1567.000000  1567.000000  1567.000000  1567.000000  1567.000000   
mean      0.162033     0.055264     0.967060     0.449571     0.836579   
std       0.035006     0.024407     0.029726     0.059854     0.069807   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.147885     0.051820     0.959176     0.411959     0.829783   
50%       0.160273     0.054698     0.969906     0.441333     0.842506   
75%       0.174238     0.057503     0.980266     0.495505     0.857767   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

        feature_20   feature_21   feature_22   feature_23   feature_24  \
count  1567.000000  1567.000000  1567.000000  1567.000000  1567.000000   
mean      0.823362     0.214255     0.738279     0.500450     0.501762   
std       0.061151     0.087610     0.080769     0.111685     0.100339   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.792108     0.170291     0.705128     0.454756     0.461082   
50%       0.826818     0.227544     0.728615     0.499281     0.509356   
75%       0.859700     0.250848     0.776923     0.536883     0.559684   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

        feature_25   feature_26   feature_27   feature_28   feature_29  \
count  1567.000000  1567.000000  1567.000000  1567.000000  1567.000000   
mean      0.870658     0.944339     0.866902     0.545897     0.597494   
std       0.128370     0.092255     0.162383     0.186973     0.143592   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.791799     0.928853     0.687751     0.431532     0.500000   
50%       0.927828     0.967703     0.948543     0.527330     0.601568   
75%       0.943231     0.975838     0.957017     0.694895     0.699234   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

        feature_30   feature_31   

## Handle Class Imbalance

In [14]:
# Extract features and target variable from data frame
targetOutcome = pd.DataFrame(df_normalized,columns=['HouseLineTestingResult'])
allInputs = pd.DataFrame(df_normalized,columns=df.columns.difference(['HouseLineTestingResult']))

In [15]:
# Modify data by handling class imbalance using SMOTE method
print('Original training targets shape {}'.format(Counter(targetOutcome.values.ravel())))
sm = SMOTE(random_state=0)
allInputs_resample, targetOutcome_resample = sm.fit_sample(allInputs, targetOutcome.values.ravel())
print('Resampled dataset shape {}'.format(Counter(targetOutcome_resample)))

Original training targets shape Counter({0.0: 1463, 1.0: 104})
Resampled dataset shape Counter({0.0: 1463, 1.0: 1463})


## Feature Selection

In [16]:
# Create linear regression model as the estimator
estimator = LinearRegression()
# From the Filter method, we select 10 features -- this number is arbitary (need to find the optimal number with further analysis)
selector = RFE(estimator, 10, step=1)
# Learn from this dataset
selector = selector.fit(allInputs, targetOutcome)
# The mask of selected features (which variables are selected)
print(selector.support_)
# Selected features are ranked 1. The variable with the highest rank is the one that is removed first
print(selector.ranking_)
# Print the name of selected features
f = selector.get_support(True) # the most important features
print('\033[1m\033[4mSeledted features from Backward Stepwise Model Selection are: \033[0m')
for f_index in f:
    print(df.columns[f_index])

[False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False  True
 False False False False False False False False False False False False
  True False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False  True False  T

In [17]:
# Create a new copy of data frame with selected features
selected_feature_names = ['feature_47','feature_60','feature_162','feature_164','feature_171','feature_196','feature_209','feature_324','feature_464','feature_466']
allInputs_feature_selected = allInputs[selected_feature_names].copy()

In [18]:
# View first 5 rows
allInputs_feature_selected.head()

feature_47  feature_60  feature_162  feature_164  feature_171  feature_196  \
0    0.798994    0.533284     0.009981     0.026417      0.21645     0.009205   
1    0.780757    0.555275     0.088877     0.063291      0.37683     0.005046   
2    0.791108    0.810180     0.056955     0.064392      0.05772     0.008469   
3    0.794558    0.747554     0.081473     0.059439      0.05504     0.010574   
4    0.529771    0.617807     0.039136     0.022014      0.11173     0.009560   

   feature_209  feature_324  feature_464  feature_466  
0          0.0     0.317877          0.0          0.0  
1          0.0     0.430930          0.0          0.0  
2          0.0     0.277554          0.0          0.0  
3          0.0     0.272055          0.0          0.0  
4          0.0     0.420716          0.0          0.0

In [19]:
# Split test(90%) and train (10%) data
X, XX, Y, YY = train_test_split(allInputs_feature_selected, targetOutcome, test_size = 0.1, random_state = 0)

In [20]:
print('Train: X=%s, Y=%s' % (X.shape, Y.shape))
print('Test: X=%s, Y=%s' % (XX.shape, YY.shape))

Train: X=(1410, 10), Y=(1410, 1)
Test: X=(157, 10), Y=(157, 1)


# Build Models

## 1. Simple Neural Networks model

In [21]:
# Use multilayer perceptron (MLP) from sklearn with solver = stochastic gradient descent
mlp = MLPClassifier(hidden_layer_sizes=(10),solver='sgd',learning_rate_init= 0.01, max_iter=500)

In [22]:
# Train the model
mlp.fit(X, Y)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=10, learning_rate='constant',
              learning_rate_init=0.01, max_fun=15000, max_iter=500,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=None, shuffle=True, solver='sgd',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

In [23]:
# Evaluate model
scores = mlp.score(XX,YY)
print("Accuracy: %.2f%%" % (scores*100))

Accuracy: 97.45%


## 2. Deep Neural Networks model (DNN)

In [24]:
# Feature columns describe how to use the input.
my_feature_columns = []
for key in selected_feature_names:
    my_feature_columns.append(tf.feature_column.numeric_column(key=key))

In [25]:
# Build a DNN with 2 hidden layers with 30 and 10 hidden nodes each.
classifier = tf.estimator.DNNClassifier(
    feature_columns=my_feature_columns,
    # Two hidden layers of 30 and 10 nodes respectively.
    hidden_units=[30, 10],
    # The model must choose between 2 classes (Success or Failed)
    n_classes=2)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_model_dir': 'C:\\Users\\salin\\AppData\\Local\\Temp\\tmprw1915lq', '_tf_random_seed': None, '_save_summary_steps': 100, '_save_checkpoints_steps': None, '_save_checkpoints_secs': 600, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_keep_checkpoint_max': 5, '_keep_checkpoint_every_n_hours': 10000, '_log_step_count_steps': 100, '_train_distribute': None, '_device_fn': None, '_protocol': None, '_eval_distribute': None, '_experimental_distribute': None, '_experimental_max_worker_delay_secs': None, '_session_creation_timeout_secs': 7200, '_service': None, '_cluster_spec': ClusterSpec({}), '_task_type': 'worker', '_task_id': 0, '_global_id_in_cluster': 0, '_master': '', '_evaluation_master': '', '_is_chief': True, '_num_ps_replicas': 0, '_num_worker_replicas': 1}


In [26]:
# Train the Model.
classifier.train(
    input_fn=lambda: input_fn(X, Y, training=True),
    steps=500)

Instructions for updating:
If using Keras pass *_constraint arguments to layers.
Instructions for updating:
Use Variable.read_value. Variables in 2.X are initialized automatically both in eager and graph (inside tf.defun) contexts.
INFO:tensorflow:Calling model_fn.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO

In [27]:
# Evaluate model
eval_result = classifier.evaluate(
    input_fn=lambda: input_fn(XX, YY, training=False))
scores = eval_result['accuracy']
print("Accuracy: %.2f%%" % (scores*100))

INFO:tensorflow:Calling model_fn.

If you intended to run this layer in float32, you can safely ignore this warning. If in doubt, this warning is likely only an issue if you are porting a TensorFlow 1.X model to TensorFlow 2.

To change all layers to have dtype float64 by default, call `tf.keras.backend.set_floatx('float64')`. To change just this layer, pass dtype='float64' to the layer constructor. If you are the author of this layer, you can disable autocasting by passing autocast=False to the base Layer constructor.

INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2020-03-15T22:52:34Z
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from C:\Users\salin\AppData\Local\Temp\tmprw1915lq\model.ckpt-500
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Inference Time : 0.54454s
INFO:tensorflow:Finished evaluation at 2020-03-15-22:52:34
INFO:tensorflow:Saving dict for global step 500: ac

## 3. Recurrent Neural Networks model (RNN)

In [28]:
# SimpleRNN model
model = Sequential()
model.add(Embedding(input_dim=10, output_dim=1))
# The output of GRU will be a 3D tensor of shape (batch_size, timesteps, 256)
model.add(GRU(256, return_sequences=True))
# The output of SimpleRNN will be a 2D tensor of shape (batch_size, 128)
model.add(SimpleRNN(128))
model.add(Dense(1))

# View inside the network
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, None, 1)           10        
_________________________________________________________________
gru (GRU)                    (None, None, 256)         198912    
_________________________________________________________________
simple_rnn (SimpleRNN)       (None, 128)               49280     
_________________________________________________________________
dense (Dense)                (None, 1)                 129       
Total params: 248,331
Trainable params: 248,331
Non-trainable params: 0
_________________________________________________________________


In [29]:
model.compile(loss='mean_squared_error', optimizer='rmsprop')

# Fit model
model.fit(X, Y, validation_data=(XX, YY), epochs=3, batch_size=64)

Train on 1410 samples, validate on 157 samples
Epoch 1/3
1410/1410 [==============================] - 2s 2ms/sample - loss: 0.0710 - val_loss: 0.0248
Epoch 2/3
1410/1410 [==============================] - 1s 401us/sample - loss: 0.0666 - val_loss: 0.0250
Epoch 3/3
1410/1410 [==============================] - 1s 407us/sample - loss: 0.0700 - val_loss: 0.0252


In [30]:
# Evaluate model
scores = model.evaluate(XX, YY, verbose=0)
print("Accuracy: %.2f%%" % (scores*100))

Accuracy: 2.52%


***
**Summary:**
1. The accuracy of test data from **Simple Neural Networks model** is **97.45%** which is very good.
2. The accuracy of test data from **Deep Neural Networks model** is also **97.45%** which is very good.
3. However, the accuracy of test data from **Recurrent Neural Networks model** is only **2.52%** which is very low.

Note: This is a very simple model without any advanced techniques. The result is just in a preliminary investigation stage and we can't make any conclusion if one of the model is out perform another model.
***

***
**Solicit Feedback:** 

- For **DNN**, I used **DNNClassifier** from tensorflow since it's the simplest one. The accuracy rate on test data is pretty high. How do we test for the overfitting?
- For **RNN**, I used **SimpleRNN** from tensorflow since it's the simplest one. I didn't include the timestamp from the label data set at all. My question is, how do we include the time as an input since RNN is about recurring.
***